Added helper functions for Graph response to client

pawanrawal opened the pull request:
1. Added Values, HasValue and NumChildren functions

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/dgraph-io/dgraph/74)
<!-- Reviewable:end -->

manishrjain commented:
Given that the client would be a library, create a new file -- call it client.go, and then add tests for client.go. Use it in the temporary main.go that you have here.

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[client/go/main.go, line 47 \[r1\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGGk55YBtzNPDym6job:-KGGk55YBtzNPDym6joc:-1884721806) ([raw file](https://github.com/dgraph-io/dgraph/blob/a69c58193335a7ee745de8dc9107759707733ce2/client/go/main.go#L47)):*
I think generally we should only have 1 value, not more. So, Call it Value(..), and return a string.

Also, add a LOG(FATAL) there if it has more than 1 valid value.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-:-KGGjcEIoA0lvsBgw2H8:253264242)*
<!-- Sent from Reviewable.io -->

pawanrawal commented:
Okay

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[client/go/main.go, line 47 \[r1\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGGk55YBtzNPDym6job:-KGGkipAXRx1vLuQYzcu:1140299661) ([raw file](https://github.com/dgraph-io/dgraph/blob/a69c58193335a7ee745de8dc9107759707733ce2/client/go/main.go#L47)):*
I may not be understanding this correctly, but values in Result is a slice of byte slices. So it could have multiple values right? For e.g. in tests `name` will have multiple values.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-:-KGGl_LiHa27ueQw6L_p:676958649)*
<!-- Sent from Reviewable.io -->

manishrjain commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[client/go/main.go, line 47 \[r1\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGGk55YBtzNPDym6job:-KGGliOD8Px5mqwf1LaU:156105343) ([raw file](https://github.com/dgraph-io/dgraph/blob/a69c58193335a7ee745de8dc9107759707733ce2/client/go/main.go#L47)):*
For a given UID, there'll only be one value. But, for a given UID, there could be multiple uids. In other words, in a posting list, we typically only have 1 value. We currently should log fatal if we have more than 1 value.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/74)*
<!-- Sent from Reviewable.io -->

pawanrawal commented:
Made the change. Also shifted files to directly inside client folder as we are only going to have the go client in this codebase. So segregation​ based on language may not make sense. For now, I have main inside go folder but that will go away at some point or the other.

---

Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[client/go/main.go, line 47 \[r1\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGGk55YBtzNPDym6job:-KGGmQ23lUaqMnwXdPbj:1006408353) ([raw file](https://github.com/dgraph-io/dgraph/blob/a69c58193335a7ee745de8dc9107759707733ce2/client/go/main.go#L47)):*
I get what you are saying here. Right now, Values returns all values corresponding to a predicate. I think I will have to modify it to return value for a  predicate+subject combination and then it will return just one value. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-:-KGIDxpQjVWvHYCv6-8a:1282691625)*
<!-- Sent from Reviewable.io -->

manishrjain commented:
Reviewed 2 of 4 files at r2, 6 of 6 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[client/client.go, line 31 \[r3\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGQ_Y0c_p6GmcCnyYZx:-KGQ_Y0d-R9829bRuDQ4:638403803) ([raw file](https://github.com/dgraph-io/dgraph/blob/c8431a63e8b94ac57d506e2af8738b189f578617/client/client.go#L31)):*
Call it indexOf instead.

---

*[client/client.go, line 47 \[r3\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGQ_mmVQo6GSBM4UB0z:-KGQ_mmVQo6GSBM4UB1-:325947511) ([raw file](https://github.com/dgraph-io/dgraph/blob/c8431a63e8b94ac57d506e2af8738b189f578617/client/client.go#L47)):*
Instead of doing this population, can you put this logic directly in the ​iteration? Would avoid having to potentially create millions of nodes.

Check if e.gr is nil

---

*[client/client.go, line 53 \[r3\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGQ_qLy_uv-qucOd_sR:-KGQ_qLy_uv-qucOd_sS:-2126650782) ([raw file](https://github.com/dgraph-io/dgraph/blob/c8431a63e8b94ac57d506e2af8738b189f578617/client/client.go#L53)):*
idx := 

Also, if idx == -1, then maybe log fatal. I think you do expect the uid to be present, right?

---

*[client/client.go, line 61 \[r3\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGQaMt8xuXa2gcDE9Ob:-KGQaMt8xuXa2gcDE9Oc:905798414) ([raw file](https://github.com/dgraph-io/dgraph/blob/c8431a63e8b94ac57d506e2af8738b189f578617/client/client.go#L61)):*
This would allocate a lot of memory. I think we can avoid it.

---

*[client/client.go, line 84 \[r3\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGQaRWTRfRgcuWvcc-u:-KGQaRWTRfRgcuWvcc-v:1689613788) ([raw file](https://github.com/dgraph-io/dgraph/blob/c8431a63e8b94ac57d506e2af8738b189f578617/client/client.go#L84)):*
Do the iteration here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/74)*
<!-- Sent from Reviewable.io -->

pawanrawal commented:
Review status: 6 of 7 files reviewed at latest revision, 5 unresolved discussions.

---

*[client/client.go, line 31 \[r3\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGQ_Y0c_p6GmcCnyYZx:-KGQrp-na3dEuXC2X1Af:-499146009) ([raw file](https://github.com/dgraph-io/dgraph/blob/c8431a63e8b94ac57d506e2af8738b189f578617/client/client.go#L31)):*
Done.

---

*[client/client.go, line 47 \[r3\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGQ_mmVQo6GSBM4UB0z:-KGQrqYkzIDFrh-Zbb7w:676635175) ([raw file](https://github.com/dgraph-io/dgraph/blob/c8431a63e8b94ac57d506e2af8738b189f578617/client/client.go#L47)):*
Done

---

*[client/client.go, line 53 \[r3\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGQ_qLy_uv-qucOd_sR:-KGQrtPvw814Iamj8W9C:676635175) ([raw file](https://github.com/dgraph-io/dgraph/blob/c8431a63e8b94ac57d506e2af8738b189f578617/client/client.go#L53)):*
Done

---

*[client/client.go, line 61 \[r3\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGQaMt8xuXa2gcDE9Ob:-KGQrukgDgije09WP0oS:-499146009) ([raw file](https://github.com/dgraph-io/dgraph/blob/c8431a63e8b94ac57d506e2af8738b189f578617/client/client.go#L61)):*
Done.

---

*[client/client.go, line 84 \[r3\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGQaRWTRfRgcuWvcc-u:-KGQrwMjF4Q64kI0PFks:-499146009) ([raw file](https://github.com/dgraph-io/dgraph/blob/c8431a63e8b94ac57d506e2af8738b189f578617/client/client.go#L84)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/74)*
<!-- Sent from Reviewable.io -->

manishrjain commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[client/client.go, line 67 \[r4\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGR1i4u-ux3lXMY29sy:-KGR1i4u-ux3lXMY29sz:-833004653) ([raw file](https://github.com/dgraph-io/dgraph/blob/39819e67f085d16355fa80a4f3977d0fb0a54938/client/client.go#L67)):*
glog.WithField("attribute", e.Attribute).Fatal("Nil graphResponse ...")

---

*[client/client.go, line 73 \[r4\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGR1o7KDVk5wq2LhZhT:-KGR1o7KDVk5wq2LhZhU:420432659) ([raw file](https://github.com/dgraph-io/dgraph/blob/39819e67f085d16355fa80a4f3977d0fb0a54938/client/client.go#L73)):*
glog.WithField(..).WithField(..) or
glog.WithFields({
  "key": val,
   "key2": val2,
}).Fatal(..)

---

*[client/client.go, line 77 \[r4\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGR22WAd-UExDvt07-6:-KGR22WAd-UExDvt07-7:-1935904561) ([raw file](https://github.com/dgraph-io/dgraph/blob/39819e67f085d16355fa80a4f3977d0fb0a54938/client/client.go#L77)):*
Do this check before doing the binary search, save some CPU cycles.

if len(grChild.Children) > 0 {
  continue
}

idx := indexOf...

---

*[client/client.go, line 81 \[r4\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGR2Nb-AEnZVE0y8aoz:-KGR2Nb-AEnZVE0y8ap-:-544671484) ([raw file](https://github.com/dgraph-io/dgraph/blob/39819e67f085d16355fa80a4f3977d0fb0a54938/client/client.go#L81)):*
Do we really need to store it? Most likely, the user will only ask for it once.

If we store it, then by the time iteration of the tree is done, we'd have allocated memory for all the nodes in the graph -- which is not particularly better than doing the preprocessing.

---

*[client/client.go, line 86 \[r4\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGR2j6iZ03LKhjVOloc:-KGR2j6jX3X26gdoxuYI:-556454711) ([raw file](https://github.com/dgraph-io/dgraph/blob/39819e67f085d16355fa80a4f3977d0fb0a54938/client/client.go#L86)):*
range returns index, val. So, k here is the index, not the value.

What you want is:
for i, k := range e.values

Then you don't have to do i++.

---

*[client/client.go, line 95 \[r4\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGR2vIHK9efve77maIk:-KGR2vIHK9efve77maIl:1202692747) ([raw file](https://github.com/dgraph-io/dgraph/blob/39819e67f085d16355fa80a4f3977d0fb0a54938/client/client.go#L95)):*
Avoid this. Just do an iterative check.

---

*[client/client.go, line 102 \[r4\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGR3-8jrO9cFlKEYTQI:-KGR3-8jrO9cFlKEYTQJ:321074007) ([raw file](https://github.com/dgraph-io/dgraph/blob/39819e67f085d16355fa80a4f3977d0fb0a54938/client/client.go#L102)):*
Iterative value return.

---

*[client/client.go, line 111 \[r4\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGR37bMLRdaicY-ubcg:-KGR37bMLRdaicY-ubch:-1845548258) ([raw file](https://github.com/dgraph-io/dgraph/blob/39819e67f085d16355fa80a4f3977d0fb0a54938/client/client.go#L111)):*
This can easily be:

if len(e.children) > 0 {
  return e.children
}

// Population logic here outside loop.

---

*[client/client.go, line 114 \[r4\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGR3H606UJ-gxXgXcie:-KGR3H61sUnrgTqHUcmX:-2097981739) ([raw file](https://github.com/dgraph-io/dgraph/blob/39819e67f085d16355fa80a4f3977d0fb0a54938/client/client.go#L114)):*
Always return, even after Fatal. This way if someone changes to Error later, the code would still be right.

Also, best thing about the logging library is WithField(..), use it. And keep Fatalf out, use Fatal instead.

---

*[client/client.go, line 116 \[r4\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGR3_kIR3mILf7OitDe:-KGR3_kJQbjdrhklFuQ0:-2126899440) ([raw file](https://github.com/dgraph-io/dgraph/blob/39819e67f085d16355fa80a4f3977d0fb0a54938/client/client.go#L116)):*
A better name might be predChild

---

*[client/client.go, line 122 \[r4\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGR3VkkGQUgcTRjM6Ut:-KGR3Vkloq47lYONfLvK:-1280722315) ([raw file](https://github.com/dgraph-io/dgraph/blob/39819e67f085d16355fa80a4f3977d0fb0a54938/client/client.go#L122)):*
return

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/74)*
<!-- Sent from Reviewable.io -->

pawanrawal commented:
Review status: 5 of 7 files reviewed at latest revision, 11 unresolved discussions.

---

*[client/client.go, line 67 \[r4\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGR1i4u-ux3lXMY29sy:-KGRPZD15Z0VjvQfYRwL:-499146009) ([raw file](https://github.com/dgraph-io/dgraph/blob/39819e67f085d16355fa80a4f3977d0fb0a54938/client/client.go#L67)):*
Done.

---

*[client/client.go, line 73 \[r4\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGR1o7KDVk5wq2LhZhT:-KGRPeU9wKLKh2GJnlJx:-499146009) ([raw file](https://github.com/dgraph-io/dgraph/blob/39819e67f085d16355fa80a4f3977d0fb0a54938/client/client.go#L73)):*
Done.

---

*[client/client.go, line 77 \[r4\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGR22WAd-UExDvt07-6:-KGRPg-2TwMflHz2m3MN:-499146009) ([raw file](https://github.com/dgraph-io/dgraph/blob/39819e67f085d16355fa80a4f3977d0fb0a54938/client/client.go#L77)):*
Done.

---

*[client/client.go, line 81 \[r4\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGR2Nb-AEnZVE0y8aoz:-KGR7zS11-ZIJLGROZHb:665607210) ([raw file](https://github.com/dgraph-io/dgraph/blob/39819e67f085d16355fa80a4f3977d0fb0a54938/client/client.go#L81)):*
Since we are optimizing for memory here, we can do without storing this.

---

*[client/client.go, line 86 \[r4\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGR2j6iZ03LKhjVOloc:-KGRPtV4Tk0HI8ih0mWc:1170231063) ([raw file](https://github.com/dgraph-io/dgraph/blob/39819e67f085d16355fa80a4f3977d0fb0a54938/client/client.go#L86)):*
Done. Added a test to verify this. My bad here.

---

*[client/client.go, line 95 \[r4\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGR2vIHK9efve77maIk:-KGRPxYgIWKX4uHV-3bz:-499146009) ([raw file](https://github.com/dgraph-io/dgraph/blob/39819e67f085d16355fa80a4f3977d0fb0a54938/client/client.go#L95)):*
Done.

---

*[client/client.go, line 102 \[r4\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGR3-8jrO9cFlKEYTQI:-KGRPzg90BflG5lBFvDq:-499146009) ([raw file](https://github.com/dgraph-io/dgraph/blob/39819e67f085d16355fa80a4f3977d0fb0a54938/client/client.go#L102)):*
Done.

---

*[client/client.go, line 111 \[r4\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGR37bMLRdaicY-ubcg:-KGRQ3haJoJZFyWK-R_d:-499146009) ([raw file](https://github.com/dgraph-io/dgraph/blob/39819e67f085d16355fa80a4f3977d0fb0a54938/client/client.go#L111)):*
Done.

---

*[client/client.go, line 114 \[r4\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGR3H606UJ-gxXgXcie:-KGRQ4j8sSwL4nl6eRey:-1926026889) ([raw file](https://github.com/dgraph-io/dgraph/blob/39819e67f085d16355fa80a4f3977d0fb0a54938/client/client.go#L114)):*
Good idea, didn't think about it this way. Thanks for sharing the idea behind this. Would returning an error make sense ?

---

*[client/client.go, line 116 \[r4\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGR3_kIR3mILf7OitDe:-KGRQK38edyfoyO-cxjz:-499146009) ([raw file](https://github.com/dgraph-io/dgraph/blob/39819e67f085d16355fa80a4f3977d0fb0a54938/client/client.go#L116)):*
Done.

---

*[client/client.go, line 122 \[r4\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGR3VkkGQUgcTRjM6Ut:-KGRQM7tztkKtHVMIcP6:-499146009) ([raw file](https://github.com/dgraph-io/dgraph/blob/39819e67f085d16355fa80a4f3977d0fb0a54938/client/client.go#L122)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/74)*
<!-- Sent from Reviewable.io -->

manishrjain commented:
Reviewed 1 of 4 files at r2, 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[client/client.go, line 31 \[r5\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGRSahKLjJtHUgJNsSX:-KGRSahLThNvueapaR0s:1649042370) ([raw file](https://github.com/dgraph-io/dgraph/blob/b72aefdc290e1fda41f434f71129ef82ef8d8084/client/client.go#L31)):*
While I look at it, I think if we were to change how we build our Protocol Buffers, we might have a better way. For e.g., we store the UID within the Result struct, which stores the value and the UIDList. Then, in the Entity, we could have an array of pointers to Result.

The other way might be to do this processing in the server itself. Generate an entity based tree from Flatbuffer.

---

*[client/client.go, line 62 \[r5\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGRU0ut_V9ykqdW8CGV:-KGRU0ut_V9ykqdW8CGW:-1041649884) ([raw file](https://github.com/dgraph-io/dgraph/blob/b72aefdc290e1fda41f434f71129ef82ef8d8084/client/client.go#L62)):*
var m []string

---

*[client/client.go, line 84 \[r5\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGRUTBfGwnYzz3zOO7D:-KGRUTBfGwnYzz3zOO7E:-1808820852) ([raw file](https://github.com/dgraph-io/dgraph/blob/b72aefdc290e1fda41f434f71129ef82ef8d8084/client/client.go#L84)):*
Note that this isn't sufficient check to see if the value exists. I think this would end up with the same complexity as Value function itself. So, let's just get rid of this function.

---

*[client/client.go, line 104 \[r5\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGRUKwnJ-7jmS_SyuOx:-KGRUKwnJ-7jmS_SyuOy:1580284957) ([raw file](https://github.com/dgraph-io/dgraph/blob/b72aefdc290e1fda41f434f71129ef82ef8d8084/client/client.go#L104)):*
Can this be nil? Some of them might not have a value right.

---

*[client/client.go, line 136 \[r5\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGRVDXPIEbKkSuedQc5:-KGRVDXPIEbKkSuedQc6:1081604702) ([raw file](https://github.com/dgraph-io/dgraph/blob/b72aefdc290e1fda41f434f71129ef82ef8d8084/client/client.go#L136)):*
I reckon if we do this in the server itself, we wouldn't even need to do this in the client. And it would also make our other drivers pretty straightforward. What do you think?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/74)*
<!-- Sent from Reviewable.io -->

pawanrawal commented:
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[client/client.go, line 62 \[r5\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGRU0ut_V9ykqdW8CGV:-KGSITZXaVQMQS-BIDWC:-371450246) ([raw file](https://github.com/dgraph-io/dgraph/blob/b72aefdc290e1fda41f434f71129ef82ef8d8084/client/client.go#L62)):*
Cool, this way memory won't be allocated unnecessarily. 

---

*[client/client.go, line 84 \[r5\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGRUTBfGwnYzz3zOO7D:-KGSHopnrMIHi3Lwnhd1:-944644812) ([raw file](https://github.com/dgraph-io/dgraph/blob/b72aefdc290e1fda41f434f71129ef82ef8d8084/client/client.go#L84)):*
Right, we could just check the Value and if it is non-empty , we can return true.

---

*[client/client.go, line 104 \[r5\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGRUKwnJ-7jmS_SyuOx:-KGSHXx_uR6zD-nSwcDl:-51572996) ([raw file](https://github.com/dgraph-io/dgraph/blob/b72aefdc290e1fda41f434f71129ef82ef8d8084/client/client.go#L104)):*
Yes, currently we store these as an empty string. So it would return an empty byte array. Though this reminds me that for such a property HasValue should be false.

---

*[client/client.go, line 136 \[r5\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGRVDXPIEbKkSuedQc5:-KGSHxhPI2rmRuWEKRa3:-2037570573) ([raw file](https://github.com/dgraph-io/dgraph/blob/b72aefdc290e1fda41f434f71129ef82ef8d8084/client/client.go#L136)):*
Yes, we could do that. +1 for making other drivers straight forward.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/74)*
<!-- Sent from Reviewable.io -->

manishrjain commented:
Reviewed 5 of 7 files at r6, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 24 unresolved discussions.

---

*[query/query.go, line 259 \[r7\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGoMNg-GTA_9Am7hfJd:-KGoMNg0fx-FMDzYeT16:1411702844) ([raw file](https://github.com/dgraph-io/dgraph/blob/87e810384c370ee95da1356ab6eee8496825e95a/query/query.go#L259)):*
You don't need this.

---

*[query/query.go, line 264 \[r7\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGoMRRxUK1cvttWZRuV:-KGoMRRxUK1cvttWZRuW:931904045) ([raw file](https://github.com/dgraph-io/dgraph/blob/87e810384c370ee95da1356ab6eee8496825e95a/query/query.go#L264)):*
You probably don't need this.

---

*[query/query.go, line 266 \[r7\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGoJbTh3pRH1OOx5DVg:-KGoJbTi6PccHMZaQeyN:2135449273) ([raw file](https://github.com/dgraph-io/dgraph/blob/87e810384c370ee95da1356ab6eee8496825e95a/query/query.go#L266)):*
What is this for?

---

*[query/query.go, line 384 \[r7\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGoJxSc1Azu4FFL6Tmi:-KGoJxSc1Azu4FFL6Tmj:884735917) ([raw file](https://github.com/dgraph-io/dgraph/blob/87e810384c370ee95da1356ab6eee8496825e95a/query/query.go#L384)):*
This should be uids, result, err := 

---

*[query/query.go, line 403 \[r7\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGoK7pdukDlyzGeA6ww:-KGoK7pdukDlyzGeA6wx:-502873835) ([raw file](https://github.com/dgraph-io/dgraph/blob/87e810384c370ee95da1356ab6eee8496825e95a/query/query.go#L403)):*
Not all leaf nodes might have a value. Does this take care of that?

---

*[query/query.go, line 409 \[r7\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGoKYHA0QP6kcEA-LVx:-KGoKYHA0QP6kcEA-LVy:-740459416) ([raw file](https://github.com/dgraph-io/dgraph/blob/87e810384c370ee95da1356ab6eee8496825e95a/query/query.go#L409)):*
This should be called uidChild or entityChild, or just entity

---

*[query/pb/graphresponse.proto, line 5 \[r7\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGoN6m7vgZ4w_o_8x_c:-KGoN6m7vgZ4w_o_8x_d:-1389770228) ([raw file](https://github.com/dgraph-io/dgraph/blob/87e810384c370ee95da1356ab6eee8496825e95a/query/pb/graphresponse.proto#L5)):*
package graph;

---

*[query/pb/graphresponse.proto, line 11 \[r7\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGoN9r9AXQmwD76tHXT:-KGoN9rA7Kxx2ogRxINx:614184394) ([raw file](https://github.com/dgraph-io/dgraph/blob/87e810384c370ee95da1356ab6eee8496825e95a/query/pb/graphresponse.proto#L11)):*
Request

---

*[query/pb/graphresponse.proto, line 16 \[r7\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGoLDWA3LPGLNxE4vnm:-KGoLDWA3LPGLNxE4vnn:-559837422) ([raw file](https://github.com/dgraph-io/dgraph/blob/87e810384c370ee95da1356ab6eee8496825e95a/query/pb/graphresponse.proto#L16)):*
bytes binary = 1;

---

*[query/pb/graphresponse.proto, line 19 \[r7\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGoN2_cxR6vj67crlUm:-KGoN2_cxR6vj67crlUn:676932775) ([raw file](https://github.com/dgraph-io/dgraph/blob/87e810384c370ee95da1356ab6eee8496825e95a/query/pb/graphresponse.proto#L19)):*
Node

---

*[query/pb/graphresponse.proto, line 20 \[r7\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGoKpycfMpZa1IdcSwp:-KGoKpycfMpZa1IdcSwq:-828431367) ([raw file](https://github.com/dgraph-io/dgraph/blob/87e810384c370ee95da1356ab6eee8496825e95a/query/pb/graphresponse.proto#L20)):*
This node should have a UID, and possibly an XID as well, which can be filled if they asked for it in the query. So, it won't be part of the values map. Probably small case.

Also, values should be called properties.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/74)*
<!-- Sent from Reviewable.io -->

pawanrawal commented:
Review status: 0 of 7 files reviewed at latest revision, 24 unresolved discussions.

---

*[query/query.go, line 259 \[r7\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGoMNg-GTA_9Am7hfJd:-KGohwe5S9HriQSgijfz:1065609439) ([raw file](https://github.com/dgraph-io/dgraph/blob/87e810384c370ee95da1356ab6eee8496825e95a/query/query.go#L259)):*
Done. My bad, this was because of my incomplete understanding of flatbuffers.

---

*[query/query.go, line 264 \[r7\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGoMRRxUK1cvttWZRuV:-KGohxYMkfaAnFCie6bW:-499146009) ([raw file](https://github.com/dgraph-io/dgraph/blob/87e810384c370ee95da1356ab6eee8496825e95a/query/query.go#L264)):*
Done.

---

*[query/query.go, line 266 \[r7\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGoJbTh3pRH1OOx5DVg:-KGoi22nnY7tjCbQc6Kt:1946852793) ([raw file](https://github.com/dgraph-io/dgraph/blob/87e810384c370ee95da1356ab6eee8496825e95a/query/query.go#L266)):*
Removed the method.

---

*[query/query.go, line 384 \[r7\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGoJxSc1Azu4FFL6Tmi:-KGoi5Bz219_vsN7ka9V:-1784343413) ([raw file](https://github.com/dgraph-io/dgraph/blob/87e810384c370ee95da1356ab6eee8496825e95a/query/query.go#L384)):*
Removed this.

---

*[query/query.go, line 403 \[r7\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGoK7pdukDlyzGeA6ww:-KGosOi3ayRmIyX9ebgH:1340695639) ([raw file](https://github.com/dgraph-io/dgraph/blob/87e810384c370ee95da1356ab6eee8496825e95a/query/query.go#L403)):*
So in case the leaf node is a value node and doesn't have a value , I store an empty byte slice. In case it's not a value node , same process of getting its children and continues in a recursive fashion.

---

*[query/query.go, line 409 \[r7\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGoKYHA0QP6kcEA-LVx:-KGosi8-UrEzMXlFS976:-921152296) ([raw file](https://github.com/dgraph-io/dgraph/blob/87e810384c370ee95da1356ab6eee8496825e95a/query/query.go#L409)):*
Calling it uidChild now.

---

*[query/pb/graphresponse.proto, line 5 \[r7\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGoN6m7vgZ4w_o_8x_c:-KGosqHyiBsJNJDdNt34:-499146009) ([raw file](https://github.com/dgraph-io/dgraph/blob/87e810384c370ee95da1356ab6eee8496825e95a/query/pb/graphresponse.proto#L5)):*
Done.

---

*[query/pb/graphresponse.proto, line 11 \[r7\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGoN9r9AXQmwD76tHXT:-KGosqfv1te_fTXQW4g6:-499146009) ([raw file](https://github.com/dgraph-io/dgraph/blob/87e810384c370ee95da1356ab6eee8496825e95a/query/pb/graphresponse.proto#L11)):*
Done.

---

*[query/pb/graphresponse.proto, line 16 \[r7\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGoLDWA3LPGLNxE4vnm:-KGosleqTFs3fqV7-6YS:-499146009) ([raw file](https://github.com/dgraph-io/dgraph/blob/87e810384c370ee95da1356ab6eee8496825e95a/query/pb/graphresponse.proto#L16)):*
Done.

---

*[query/pb/graphresponse.proto, line 19 \[r7\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGoN2_cxR6vj67crlUm:-KGospuj7EIGjZS12_Eh:-499146009) ([raw file](https://github.com/dgraph-io/dgraph/blob/87e810384c370ee95da1356ab6eee8496825e95a/query/pb/graphresponse.proto#L19)):*
Done.

---

*[query/pb/graphresponse.proto, line 20 \[r7\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGoKpycfMpZa1IdcSwp:-KGoslxIKHFauW_ME_pE:-1294701540) ([raw file](https://github.com/dgraph-io/dgraph/blob/87e810384c370ee95da1356ab6eee8496825e95a/query/pb/graphresponse.proto#L20)):*
Has UID, should I store XID too ?

---

*[client/client.go, line 31 \[r5\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGRSahKLjJtHUgJNsSX:-KGostzHqtYiuAMpO2wA:640020970) ([raw file](https://github.com/dgraph-io/dgraph/blob/b72aefdc290e1fda41f434f71129ef82ef8d8084/client/client.go#L31)):*
Processing this in server.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/74)*
<!-- Sent from Reviewable.io -->

manishrjain commented:
Reviewed 8 of 8 files at r8.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[query/query.go, line 274 \[r8\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGou10533huGhOAp4j3:-KGou106y6nzCjm93pbf:-1883923421) ([raw file](https://github.com/dgraph-io/dgraph/blob/5a62b48a117b24ec043e1ae5c9edad996f6df32c/query/query.go#L274)):*
func (g *SubGraph) preTraverse(uid uint64, dst *graph.Node)

Pass in pointer to the pb.

---

*[query/query.go, line 313 \[r8\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGowbFWNH6Im0KTcKTY:-KGowbFWNH6Im0KTcKTZ:-716242711) ([raw file](https://github.com/dgraph-io/dgraph/blob/5a62b48a117b24ec043e1ae5c9edad996f6df32c/query/query.go#L313)):*
uc := new(graph.Node)

var err error
if uc.Properties, uc.Children, err = preTraverse(..); err != nil {
}

---

*[query/query.go, line 340 \[r8\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGowzto3Zyo2wnfYP3K:-KGowzto3Zyo2wnfYP3L:1277080095) ([raw file](https://github.com/dgraph-io/dgraph/blob/5a62b48a117b24ec043e1ae5c9edad996f6df32c/query/query.go#L340)):*
if values["_xid_"] => Assign it as a field.

---

*[query/query_test.go, line 334 \[r8\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGoxVEoreGlYvWNSPHZ:-KGoxVEpTbp5Z1hs467A:-473947610) ([raw file](https://github.com/dgraph-io/dgraph/blob/5a62b48a117b24ec043e1ae5c9edad996f6df32c/query/query_test.go#L334)):*
_xid_

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/74)*
<!-- Sent from Reviewable.io -->

pawanrawal commented:
Review status: 3 of 7 files reviewed at latest revision, 7 unresolved discussions.

---

*[query/query.go, line 274 \[r8\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGou10533huGhOAp4j3:-KGp42bNUqSzjsWgw-Bq:-499146009) ([raw file](https://github.com/dgraph-io/dgraph/blob/5a62b48a117b24ec043e1ae5c9edad996f6df32c/query/query.go#L274)):*
Done.

---

*[query/query.go, line 313 \[r8\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGowbFWNH6Im0KTcKTY:-KGp40TCOeK0oImFRpwV:-499146009) ([raw file](https://github.com/dgraph-io/dgraph/blob/5a62b48a117b24ec043e1ae5c9edad996f6df32c/query/query.go#L313)):*
Done.

---

*[query/query.go, line 340 \[r8\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGowzto3Zyo2wnfYP3K:-KGp44F4uxyuPUMtbx3d:-499146009) ([raw file](https://github.com/dgraph-io/dgraph/blob/5a62b48a117b24ec043e1ae5c9edad996f6df32c/query/query.go#L340)):*
Done.

---

*[query/query_test.go, line 334 \[r8\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGoxVEoreGlYvWNSPHZ:-KGp45NL9bSJnAOIiyPI:-499146009) ([raw file](https://github.com/dgraph-io/dgraph/blob/5a62b48a117b24ec043e1ae5c9edad996f6df32c/query/query_test.go#L334)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/74)*
<!-- Sent from Reviewable.io -->

manishrjain commented:
Review status: 3 of 7 files reviewed at latest revision, 4 unresolved discussions.

---

*[query/query.go, line 339 \[r9\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGp5vOhv76yHKy0FR4_:-KGp5vOizhvifp8aihsg:346313050) ([raw file](https://github.com/dgraph-io/dgraph/blob/7be0e39c5deaa3751eb9a054c25ed337cae0a28b/query/query.go#L339)):*
`_xid_`, with the underscores.

---

*[query/query.go, line 340 \[r9\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGp624auaGtBI5NwQik:-KGp624auaGtBI5NwQil:821157073) ([raw file](https://github.com/dgraph-io/dgraph/blob/7be0e39c5deaa3751eb9a054c25ed337cae0a28b/query/query.go#L340)):*
Also remove the property from the map.

delete(properties, "_xid_")

---

*[query/query_test.go, line 334 \[r8\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGoxVEoreGlYvWNSPHZ:-KGp6EyKC8iKtdPaRw10:1588817969) ([raw file](https://github.com/dgraph-io/dgraph/blob/5a62b48a117b24ec043e1ae5c9edad996f6df32c/query/query_test.go#L334)):*
It's `_xid_`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/74)*
<!-- Sent from Reviewable.io -->

pawanrawal commented:
Review status: 3 of 7 files reviewed at latest revision, 4 unresolved discussions.

---

*[query/query.go, line 339 \[r9\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGp5vOhv76yHKy0FR4_:-KGp9OSryZroM-XEoz8G:-499146009) ([raw file](https://github.com/dgraph-io/dgraph/blob/7be0e39c5deaa3751eb9a054c25ed337cae0a28b/query/query.go#L339)):*
Done.

---

*[query/query.go, line 340 \[r9\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGp624auaGtBI5NwQik:-KGp9P5iVCitcAbw5mCs:-499146009) ([raw file](https://github.com/dgraph-io/dgraph/blob/7be0e39c5deaa3751eb9a054c25ed337cae0a28b/query/query.go#L340)):*
Done.

---

*[query/query_test.go, line 334 \[r8\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGoxVEoreGlYvWNSPHZ:-KGp9T9rMwo6wnTZ8037:-499146009) ([raw file](https://github.com/dgraph-io/dgraph/blob/5a62b48a117b24ec043e1ae5c9edad996f6df32c/query/query_test.go#L334)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/74)*
<!-- Sent from Reviewable.io -->

manishrjain commented:
Reviewed 1 of 8 files at r8, 2 of 4 files at r9, 2 of 2 files at r10.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[query/query.go, line 335 \[r10\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGpMyZTuAzeoLwEANcH:-KGpMyZTuAzeoLwEANcI:1959314741) ([raw file](https://github.com/dgraph-io/dgraph/blob/3ee878454bc9513956c9a930a11abd20501a3a7f/query/query.go#L335)):*
This looks wrong. If we are assuming string, we should just store it like so.

Given everything is a string right now, why not just call it that instead of binary; and use a string.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/74)*
<!-- Sent from Reviewable.io -->

pawanrawal commented:
Review status: 3 of 7 files reviewed at latest revision, 2 unresolved discussions.

---

*[query/query.go, line 335 \[r10\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGpMyZTuAzeoLwEANcH:-KGpRnBe4z-R7dxEVH2F:-499146009) ([raw file](https://github.com/dgraph-io/dgraph/blob/3ee878454bc9513956c9a930a11abd20501a3a7f/query/query.go#L335)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/74)*
<!-- Sent from Reviewable.io -->

manishrjain commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Looks great! One small comment, so just fix that before merging.

---

Reviewed 1 of 8 files at r8, 4 of 4 files at r11.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[query/query_test.go, line 381 \[r11\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGpSW06scYtI-aAvFSE:-KGpSW07BLjXD5sQLxn1:-898113107) ([raw file](https://github.com/dgraph-io/dgraph/blob/8cfabdf6082e38a5a4dc0adb7c5ef09e14b252ee/query/query_test.go#L381)):*
string(...) isn't required.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-:-KGpSaZo31j6UlVOsLID:1568274667)*
<!-- Sent from Reviewable.io -->

manishrjain commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r12.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-:-KGpTteUvwrxe2RXU91m:1417088721)*
<!-- Sent from Reviewable.io -->

pawanrawal commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[query/query_test.go, line 381 \[r11\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/74#-KGpSW06scYtI-aAvFSE:-KGpU0r9gTO-pyh9N1Td:-499146009) ([raw file](https://github.com/dgraph-io/dgraph/blob/8cfabdf6082e38a5a4dc0adb7c5ef09e14b252ee/query/query_test.go#L381)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/74)*
<!-- Sent from Reviewable.io -->
